cpp/include/raft/core/device_resources_snmg.hpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_resources.hpp>
+
+#include <nccl.h>
+#include <omp.h>
+
+#include <memory>
+#include <vector>
+
+/**
+ * @brief Error checking macro for NCCL runtime API functions.
+ *
+ * Invokes a NCCL runtime API function call, if the call does not return ncclSuccess, throws an
+ * exception detailing the NCCL error that occurred
+ */
+#define RAFT_NCCL_TRY(call)                        \
+  do {                                             \
+    ncclResult_t const status = (call);            \
+    if (ncclSuccess != status) {                   \
+      std::string msg{};                           \
+      SET_ERROR_MSG(msg,                           \
+                    "NCCL error encountered at: ", \
+                    "call='%s', Reason=%d:%s",     \
+                    #call,                         \
+                    status,                        \
+                    ncclGetErrorString(status));   \
+      throw raft::logic_error(msg);                \
+    }                                              \
+  } while (0);
+
+namespace raft {
+
+/**
+ * @brief SNMG resource container object that stores a NCCL clique and
+ * all necessary resources used for calling device functions, cuda kernels,
+ * libraries and/or NCCL communications on each GPU
+ *
+ * Note that the NCCL communications sometimes require
+ * the use of multiple threads to avoid hangs
+ */
+class device_resources_snmg : public device_resources {
+ public:
+  /**
+   * @brief Construct a SNMG resources instance with all available GPUs
+   */
+  device_resources_snmg() : device_resources(), root_rank_(0)
+  {
+    cudaGetDevice(&main_gpu_id_);
```

Comment:
We should probably document this as the expected behavior or allow the user to set the id of the gpu that should be used as the "default" just in case they would like it to be different than one of the GPUs in the nccl clique for some reason.

Reply:
Expanded the documentation on that point. The default GPU is the GPU currently in use at the time the constructor is called. It could theoretically not be part of the clique. However, I worry a bit that allowing the user to change the default GPU through a constructor parameter might cause the `device_resources` constructor to be first called on the wrong GPU and then be associated with an other GPU which might cause issues down the line. Also, in term of consistency the `device_resources` object is always associated to the GPU it was instantiated with.